File: leetcode/js/longest_substring.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const chars = {};
    let len = 0;
    let j = 0;

    for (let i = 0; i < s.length; i++) {
        let charpos  = chars?.[s[i]] || 0;

        j = Math.max(j, charpos);
        len = Math.max(len, i - j + 1);
        chars[s[i]] = i + 1;
    }

    return len;
};

File: leetcode/java/LongestSubstring.java
import java.util.HashMap;

class LongestSubstring {
    public int lengthOfLongestSubstring(String s) {
        final HashMap<Character, Integer> chars = new HashMap<Character, Integer>();
        int j = 0;
        int len = 0;

        for (int i = 0; i < s.length(); i++) {
            int charpos = chars.getOrDefault(s.charAt(i), 0);

            j = Math.max(charpos, j);
            len = Math.max(len, i - j + 1);
            chars.put(s.charAt(i), i + 1);
        }

        return len;
    }
}
